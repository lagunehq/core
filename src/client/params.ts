import { AccountField, AccountSource } from '../entities/account';
import { FilterContext } from '../entities/filter';
import { NotificationType } from '../entities/notification';
import { OAuthClient } from '../entities/oauth';
import { PushSubscriptionAlerts } from '../entities/push-subscription';
import { StatusVisibility } from '../entities/status';

export interface LoginParams {
  /** URI of the instance */
  uri: string;
  /** Token of your account */
  accessToken?: string;
}

export interface PaginationParams {
  /** Get a list of items with ID less than this value */
  max_id?: string | null;
  /** Get a list of items with ID greater than this value including this ID */
  since_id?: string | null;
  /** Get a list of items with ID greater than this value exluding this ID */
  min_id?: string | null;
  /** Maximum number of items to get */
  limit?: number | null;
}

export interface UpdateCredentialsParams {
  /** Display name */
  display_name?: string | null;
  /** Biography */
  note?: string | null;
  /** Avatar encoded using `multipart/form-data` */
  avatar?: any | null;
  /** Header image encoded using `multipart/form-data` */
  header?: any | null;
  /** Enable follow requests */
  locked?: boolean | null;
  /**
   * privacy: Default post privacy preference
   * sensitive: Whether to mark statuses as sensitive by default
   * language: Override language on statuses by default (ISO6391)
   */
  source?: Pick<AccountSource, 'privacy' | 'sensitive' | 'language'> | null;
  /** Profile metadata (max. 4) */
  fields_attributes?:
    | [AccountField]
    | [AccountField, AccountField]
    | [AccountField, AccountField, AccountField]
    | [AccountField, AccountField, AccountField, AccountField]
    | null;
}

export interface CreateAccountParams {
  /** Username to create */
  username: string;
  /** Password of the user */
  password: string;
  /** Email of the user */
  email: string;
  /** Whether the user has been agreed to the agreement of the Mastodon instance */
  agreement: boolean;
}

export interface ReportAccountParams {
  /** The ID of the account to report */
  account_id: string;
  /** The IDs of statuses to report as array */
  status_ids?: string[] | null;
  /** Reason for the report (up to 1,000 characters) */
  comment?: string | null;
  /** Whether to forward to the remote admin (in case of a remote account) */
  forward?: boolean | null;
}

export interface CreateAppParams {
  /** Name of your application */
  client_name: string;
  /** Where the user should be redirected after authorization */
  redirect_uris: string;
  /** Space separated list of scopes */
  scopes: string;
  /** URL to the homepage of your app */
  website?: string | null;
}

export type GrantType = 'authorization_code' | 'password';

export type FetchAccessTokenParams<
  T extends GrantType | undefined = undefined
> = {
  /** Grant type */
  grant_type: T;
} & (T extends 'authorization_code'
  ? OAuthClient & {
      /** Authorization code */
      code: string;

      /** Redirect URI which used for the authorization */
      redirect_uri: string;
    }
  : T extends 'password'
  ? {
      /** Password */
      password: string;

      /** Username */
      username: string;
    }
  : never);

export interface ModifyMediaAttachmentParams {
  /** Media to be uploaded (encoded using `multipart/form-data`) */
  file: any;
  /** A plain-text description of the media, for accessibility (max 420 chars) */
  descriptions?: string | null;
  /** Focal point: Two floating points, comma-delimited */
  focus?: string | null;
}

export interface ModifyFilterParams {
  /** String that contains keyword or phrase */
  phrase?: string | null;
  /** Array of strings that means filtering context. each string is one of `home`, `notifications`, `public`, `thread`. At least one context must be specified */
  context?: FilterContext[] | null;
  /** Filtered toots will disappear irreversibly, even if filter is later removed */
  irreversible?: boolean | null;
  /** Boolean that indicates word match. */
  whole_word?: boolean | null;
  /** The simestamp for expire time */
  expires_in?: number | null;
}

export interface ModifyListParams {
  /** Title of the list */
  title: string;
}

export interface ModifyListAccountsParams {
  /** Array of account IDs */
  account_ids: string[];
}

export interface FetchNotificationsParams extends PaginationParams {
  /** Array of notifications to exclude (Allowed values: "follow", "favourite", "reblog", "mention") */
  exclude_types?: NotificationType[] | null;
}

export interface AddPushSubscriptionParams {
  subscription: {
    /** Endpoint URL that called when notification is happen. */
    endpoint: string;

    keys: {
      /** User agent public key. Base64 encoded string of public key of ECDH key that using 'prime256v1' curve. */
      p256dh: string;
      /** Auth secret. Base64 encoded string of 16 bytes random data. */
      auth: string;
    };
  };
  data?: {
    alerts?: Partial<PushSubscriptionAlerts> | null;
  } | null;
}

export type UpdatePushSubscriptionParams = Pick<
  AddPushSubscriptionParams,
  'data'
>;

export interface FollowAccountParams {
  /** Whether the followed accountâ€™s reblogs will show up in the home timeline */
  reblogs?: boolean | null;
}

export interface MuteAccountParams {
  /** Whether the mute will mute notifications or not */
  notifications: boolean;
}

export interface SearchParams {
  /** The search query */
  q: string;
  /** Attempt WebFinger look-up */
  resolve?: boolean | null;
}

export interface SearchAccountsParams extends SearchParams {
  /** Maximum number of matching accounts to return (default: `40`) */
  limit?: number | null;
  /** Only who the user is following */
  following?: boolean | null;
}

export type CreateStatusParams<
  MediaIds extends string[] | null | undefined = string[]
> = {
  /** local ID of the status you want to reply to */
  in_reply_to_id?: string | null;
  /** Array of media IDs to attach to the status (maximum 4) */
  media_ids?: MediaIds | null;
  /** Set this to mark the media of the status as NSFW */
  sensitive?: boolean | null;
  /** Text to be shown as a warning before the actual content */
  spoiler_text?: string | null;
  /** Either "direct", "private", "unlisted" or "public" */
  visibility?: StatusVisibility | null;
  /** ISO 639-2 language code of the toot, to skip automatic detection */
  language?: string | null;
} & (MediaIds extends string[]
  ? {
      /** Text of the status */
      status?: string | null;
    }
  : {
      /** Text of the status */
      status: string;
    });

export interface FetchTimelineParams extends PaginationParams {
  /** Only return statuses originating from this instance (public and tag timelines only) */
  local?: boolean | null;
  /** Only return statuses that have media attachments */
  only_media?: boolean | null;
}

export interface FetchAccountStatusesParams extends PaginationParams {
  /** Only return statuses that have media attachments */
  only_media?: boolean | null;
  /** Only return statuses that have been pinned */
  pinned?: boolean | null;
  /** Skip statuses that reply to other statuses */
  exclude_replies?: boolean | null;
}
