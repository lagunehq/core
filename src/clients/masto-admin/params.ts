export interface AdminFetchAccountParams {
  local?: boolean | null;
  remote?: boolean | null;
  by_domain?: string | null;
  active?: boolean | null;
  pending?: boolean | null;
  disabled?: boolean | null;
  silenced?: boolean | null;
  suspended?: boolean | null;
  usename?: string | null;
  display_name?: string | null;
  email?: string | null;
  ip?: string | null;
  staff?: boolean | null;
}

export interface AdminFetchReportsParams {
  resolved?: boolean | null;
  account_id?: string | null;
  target_account_id?: string | null;
}

export type AccountActionType = 'none' | 'disable' | 'silence' | 'suspend';

export interface AdminActionAccountParams {
  type: AccountActionType;
  report_id: string;
  warning_preset_id?: string | null;
  text?: string | null;
  send_email_notification?: boolean | null;
}
