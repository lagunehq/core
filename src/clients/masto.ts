import { V1, V2 } from '../api';
import type { MastoConfig } from '../config';
import type { Http } from '../http';
import type { Ws } from '../ws';

export class MastoClient {
  readonly v1: V1.Repository;
  readonly v2: V2.Repository;

  constructor(
    readonly http: Http,
    readonly ws: Ws,
    readonly config: MastoConfig,
  ) {
    this.v1 = new V1.Repository(http, ws, config);
    this.v2 = new V2.Repository(http, config);
  }
}
