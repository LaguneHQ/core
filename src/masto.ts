import {
  AccountRepository,
  AnnouncementRepository,
  AppRepository,
  BlockRepository,
  BookmarkRepository,
  ConversationRepository,
  CustomEmojiRepository,
  DomainBlockRepository,
  EndorsementRepository,
  FavouriteRepository,
  FeaturedTagRepository,
  FilterRepository,
  FollowRequestRepository,
  InstanceRepository,
  ListRepository,
  MarkerRepository,
  MediaAttachmentRepository,
  MuteRepository,
  NotificationsRepository,
  PollRepository,
  PreferenceRepository,
  PushSubscriptionsRepository,
  ReportRepository,
  ScheduledStatusesRepository,
  StatusRepository,
  SuggestionRepository,
  TimelinesRepository,
  TrendRepository,
} from './aggregations';
import { DirectoryRepository } from './aggregations/directory';
import { version } from './decorators';
import { Results } from './entities';
import { Http } from './http';
import { Paginator } from './paginator';
import { DefaultPaginationParams } from './repository';

export type SearchType = 'accounts' | 'hashtags' | 'statuses';

export interface SearchParams extends DefaultPaginationParams {
  /** Attempt WebFinger lookup. Defaults to false. */
  readonly q: string;
  /** Enum(accounts, hashtags, statuses) */
  readonly type?: SearchType | null;
  /** Attempt WebFinger look-up */
  readonly resolve?: boolean | null;
  /** If provided, statuses returned will be authored only by this account */
  readonly accountId?: string | null;
  /** Filter out unreviewed tags? Defaults to false. Use true when trying to find trending tags. */
  readonly excludeUnreviewed?: boolean | null;
  /** Only include accounts that the user is following. Defaults to false. */
  readonly following?: boolean | null;
}

export class FacadeRepositories {
  static async login() {
    const http = ({} as unknown) as Http;
    const version = '1.0.0';
    return new FacadeRepositories(http, version);
  }

  private constructor(readonly http: Http, readonly version: string) {}

  readonly accounts = new AccountRepository(this.http, this.version);

  readonly announcements = new AnnouncementRepository(this.http, this.version);

  readonly apps = new AppRepository(this.http, this.version);

  readonly blocks = new BlockRepository(this.http, this.version);

  readonly bookmarks = new BookmarkRepository(this.http, this.version);

  readonly conversations = new ConversationRepository(this.http, this.version);

  readonly customEmojis = new CustomEmojiRepository(this.http, this.version);

  readonly directory = new DirectoryRepository(this.http, this.version);

  readonly domainBlocks = new DomainBlockRepository(this.http, this.version);

  readonly endorsements = new EndorsementRepository(this.http, this.version);

  readonly favourites = new FavouriteRepository(this.http, this.version);

  readonly featuredTags = new FeaturedTagRepository(this.http, this.version);

  readonly filters = new FilterRepository(this.http, this.version);

  readonly followRequests = new FollowRequestRepository(
    this.http,
    this.version,
  );

  readonly instances = new InstanceRepository(this.http, this.version);

  readonly lists = new ListRepository(this.http, this.version);

  readonly markers = new MarkerRepository(this.http, this.version);

  readonly mediaAttachments = new MediaAttachmentRepository(
    this.http,
    this.version,
  );

  readonly mutes = new MuteRepository(this.http, this.version);

  readonly notifications = new NotificationsRepository(this.http, this.version);

  readonly poll = new PollRepository(this.http, this.version);

  readonly preferences = new PreferenceRepository(this.http, this.version);

  readonly pushSubscriptions = new PushSubscriptionsRepository(
    this.http,
    this.version,
  );

  readonly reports = new ReportRepository(this.http, this.version);

  readonly scheduledStatuses = new ScheduledStatusesRepository(
    this.http,
    this.version,
  );

  readonly statuses = new StatusRepository(this.http, this.version);

  readonly suggestions = new SuggestionRepository(this.http, this.version);

  readonly timelines = new TimelinesRepository(this.http, this.version);

  readonly trends = new TrendRepository(this.http, this.version);

  /**
   * Search results
   * @param params Parameters
   * @return Results
   * @see https://docs.joinmastodon.org/methods/search/
   */
  @version({ since: '2.4.1' })
  search(params: SearchParams): AsyncIterable<Results> {
    return new Paginator<typeof params, Results>(
      this.http,
      `/api/v2/search`,
      params,
    );
  }
}
