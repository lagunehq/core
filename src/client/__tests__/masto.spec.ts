// tslint:disable
import axios from 'axios';
import { Masto } from '../masto';
import { FetchAccessTokenParams } from '../params';

jest.mock('axios');

describe('Masto', () => {
  let masto!: Masto;

  beforeAll(async done => {
    (axios.request as jest.Mock).mockResolvedValue({
      data: {
        version: '99.99.9',
        urls: {
          streaming_api: 'wss://example.com',
        },
      },
    });

    masto = await Masto.login({
      uri: 'https://example.com',
      accessToken: 'token',
    });

    done();
  });

  beforeEach(() => {
    (axios.request as jest.Mock).mockReset();
  });

  test('fetchAccessToken', async () => {
    const params: FetchAccessTokenParams = {
      grant_type: 'authorization_code',
      client_id: '123123123',
      client_secret: '456456456',
      code: '789789789',
      redirect_uri: 'https://example.com',
    };

    await masto.fetchAccessToken(params);
    expect(axios.request).toMatchSnapshot();
  });

  test('revokeAccessToken', async () => {
    await masto.revokeAccessToken({
      client_id: '123123123',
      client_secret: '456456456',
    });
    expect(axios.request).toMatchSnapshot();
  });

  test('fetchAccount', async () => {
    await masto.fetchAccount('123123123');
    expect(axios.request).toMatchSnapshot();
  });

  test('createAccount', async () => {
    await masto.createAccount({
      username: 'username',
      password: 'password',
      email: 'example@example.com',
      agreement: true,
    });
    expect(axios.request).toMatchSnapshot();
  });

  test('verifyCredentials', async () => {
    await masto.verifyCredentials();
    expect(axios.request).toMatchSnapshot();
  });

  // test('updateCredentials', async () => {
  //   await masto.updateCredentials({
  //     display_name: 'Neetshin',
  //     note: 'web frontend engineer',
  //     locked: false,
  //     source: {
  //       privacy: 'direct',
  //       sensitive: false,
  //       language: 'ja',
  //     },
  //     fields_attributes: [
  //       {
  //         name: 'GitHub',
  //         value: 'https://github.com/neet',
  //       },
  //     ],
  //   });
  //   expect(axios.request).toMatchSnapshot();
  // });

  test('fetchAccountFollowers', async () => {
    (await masto.fetchAccountFollowers('123123123', {
      max_id: '5',
      since_id: '3',
      min_id: '2',
      limit: 10,
    })).next();
    expect(axios.request).toMatchSnapshot();
  });

  test('fetchAccountFollowing', async () => {
    (await masto.fetchAccountFollowing('123123123', {
      max_id: '5',
      since_id: '3',
      min_id: '2',
      limit: 10,
    })).next();
    expect(axios.request).toMatchSnapshot();
  });

  test('fetchAccountStatuses', async () => {
    (await masto.fetchAccountStatuses('123123123', {
      max_id: '5',
      since_id: '3',
      min_id: '2',
      limit: 10,
    })).next();
    expect(axios.request).toMatchSnapshot();
  });

  test('followAccount', async () => {
    await masto.followAccount('123123', {
      reblogs: false,
    });
    expect(axios.request).toMatchSnapshot();
  });
});
