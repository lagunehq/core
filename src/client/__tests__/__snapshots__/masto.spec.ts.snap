// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Masto createAccount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{\\"username\\":\\"username\\",\\"password\\":\\"password\\",\\"email\\":\\"example@example.com\\",\\"agreement\\":true}",
        "headers": Object {
          "Authorization": "Bearer token",
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/accounts",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Masto fetchAccessToken 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{\\"grant_type\\":\\"authorization_code\\",\\"client_id\\":\\"123123123\\",\\"client_secret\\":\\"456456456\\",\\"code\\":\\"789789789\\",\\"redirect_uri\\":\\"https://example.com\\"}",
        "headers": Object {
          "Authorization": "Bearer token",
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/oauth/token",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Masto fetchAccount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Authorization": "Bearer token",
          "Content-Type": "application/json",
        },
        "method": "GET",
        "params": Object {},
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/accounts/123123123",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Masto fetchAccountFollowers 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Authorization": "Bearer token",
          "Content-Type": "application/json",
        },
        "method": "GET",
        "params": Object {
          "limit": 10,
          "max_id": "5",
          "min_id": "2",
          "since_id": "3",
        },
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/accounts/123123123/followers",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Masto fetchAccountFollowing 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Authorization": "Bearer token",
          "Content-Type": "application/json",
        },
        "method": "GET",
        "params": Object {
          "limit": 10,
          "max_id": "5",
          "min_id": "2",
          "since_id": "3",
        },
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/accounts/123123123/following",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Masto fetchAccountStatuses 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Authorization": "Bearer token",
          "Content-Type": "application/json",
        },
        "method": "GET",
        "params": Object {
          "limit": 10,
          "max_id": "5",
          "min_id": "2",
          "since_id": "3",
        },
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/accounts/123123123/statuses",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Masto followAccount 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{\\"reblogs\\":false}",
        "headers": Object {
          "Authorization": "Bearer token",
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/accounts/123123/follow",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Masto revokeAccessToken 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{\\"client_id\\":\\"123123123\\",\\"client_secret\\":\\"456456456\\"}",
        "headers": Object {
          "Authorization": "Bearer token",
          "Content-Type": "application/json",
        },
        "method": "POST",
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/oauth/revoke",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Masto verifyCredentials 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "{}",
        "headers": Object {
          "Authorization": "Bearer token",
          "Content-Type": "application/json",
        },
        "method": "GET",
        "params": Object {},
        "transformResponse": Array [
          [Function],
        ],
        "url": "https://example.com/api/v1/accounts/verify_credentials",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
